### ISSUE TITLE
Static Public Key Across Zites

### ISSUE THREAD

**Issue by d14na** [HelloZeroNet/ZeroNet#2217]
**Is your feature request related to a problem? Please describe.**
What is the purpose of making a user's `public key` UNIQUE on every zite, but the `auth_address` _(aka Bitcoin address)_ the SAME on every zite?

**Describe the solution you'd like**
A STATIC `public key` (across all zites) that is derived from the user's `master key`, same as their `auth_address` _(aka Bitcoin address)_.

**Describe alternatives you've considered**
1. Could utilize the deprecated `auth_key`. (btw, why was this deprecated?)
2. Create a NEW plugin to derive a STATIC `public key` from the `master key`.

**Additional context**

## So Why Is This A Problem?

Dynamic zites work very well today, by allowing users to write data to their own folder/namespace within the zite's sqlite data storage.

But what if you're NOT using Zeronet's sqlite data storage. It makes using "legacy" client/server features within zites cumbersome; since you have to rely on "legacy" authorization techniques for the user. Rather, a Zeronet-enhanced APP SERVER "should" be able to retrieve a user's `public key` from a registry (based on their cert id) and then send a challenge that can ONLY be solved by the `private key` holder. This would be AUTOMATIC, upon arriving at a ANY zite, making user authorization __BUILT-IN__ for ALL developers.

eg. It should appeal to developers wanting to bridge an existing Clearnet application over to Zeronet. In which case, it would be ideal to utilize current backend systems while making the migration to a fully decentralized solution.

**Comment by purplesyringa** [comment]
> What is the purpose of making a user's public key UNIQUE on every zite

That's not to allow sites to deanonymize you if you don't want that.

> but the auth_address (aka Bitcoin address) the SAME on every zite?

auth_address is *not* the same on every zite. The only way to make it the same is to switch your identity from "No certificate" (which used to be "Unique to this site" before, by the way) to some ID (i.e. ZeroID or KxoID, etc.). Thus, sites can't deanonymize you, but you can choose to use a single ID (*or* different IDs, which is not supported now but should be; you can make it work right now by using several domains though) for centralization.

> But what if you're NOT using Zeronet's sqlite data storage

How is SQLite even connected to what you're talking about? I guess you're just talking about file-based storage.

> since you have to rely on "legacy" authorization techniques for the user

Please explain what's legacy. If you're talking about the auth provider stuff, I doubt it's legacy.

> Rather, a Zeronet-enhanced APP SERVER "should" be able to retrieve a user's public key from a registry (based on their cert id) and then send a challenge that can ONLY be solved by the private key holder.

Ok, so are we talking about user verification? In this case:

1. Automatically signing everything *must not* be even considered. Signing lots of data is insecure, eats CPU resources and so on. Instead, an app should decide what (and whether) to sign data, which is currently possible with PeerMessage (you can add BackgroundProcessing for performance and reliability).
2. There's a (fairly obvious) way to get current user public key which you can then share or save or do whatever you want. There's also API to sign and verify data. Can't you use that?

> making user authorization BUILT-IN for ALL developers.

What *is* built-in user authorization? We can't just allow *all* data to be saved. The site owner has to control this. Now, you can either use built-in file share mechanisms or use PeerMessage to simulate whatever you want.

> It should appeal to developers wanting to bridge an existing Clearnet application over to Zeronet

Like what?

> In which case, it would be ideal to utilize current backend systems while making the migration to a fully decentralized solution.

How can this even be possible? Are you talking about bridging as in "make all decentralized data decentralized and leave all centralized data centralized, but also use a central service to bridge these two storages"? If that's the case, you can already do that in an obvious way.

> Could utilize the deprecated auth_key. (btw, why was this deprecated?)

I guess that's because it was centralized and allowed fingerprinting. Also, it wasn't really secure because all sites could use it to sign (possibly dangerous) commands.

> Create a NEW plugin to derive a STATIC public key from the master key.

Again, what for? Do you *really* want a static public key? You can make a static *address* currently (either use existing solutions like ZeroID or generate the IDs you need on the fly ~~like in my upcoming project~~). If you only need to sign or verify data, you don't need public keys -- addresses are enough. If you want to encrypt or decrypt data, that's a bit more of a problem, but you can simply use different encryption keys for different sites, which should be better for security, by the way.

**Comment by d14na** [comment]
> That's not to allow sites to deanonymize you if you don't want that.

ok

> auth_address is not the same on every zite.

yes. what I meant was, while using a cert id

> How is SQLite even connected to what you're talking about?

just in terms of how the code retrieves user data

> Please explain what's legacy

i consider username/password auth to be legacy (as in broken, and can't be fixed). p2p web SHOULD be using public key crypto

> There's a (fairly obvious) way to get current user public key which you can then share or save or do whatever you want. There's also API to sign and verify data. Can't you use that?

yes, this is my point. this works fine, however, it doesn't work cross-zite (for the reason you stated in your first response).

## My Very Specific Use-Case

(I'll skip the details, but this is for supporting a new [Certificate Id](https://0net.io/nametag.bit) service)

1. User registers for a new id (zeroguest@nametag.bit).
2. Their public key (eg. `A//iXrqNyBfGY6j8uwV7vaiZ+GRmLdzmA6AjKq+VAzbf`) will be stored in a PUBLIC registry; similar to the [MIT Key Server](http://pgp.mit.edu/).
3. Existing client/server app will migrate their FRONT-END to Zeronet, leaving the SERVER as-is.
4. The new Zeronet frontend SHOULD support PUBLIC KEY authentication, NOT username/password.

## Why Do This?

To encourage more existing applications to move and/or mirror to the Zeronet. There are numerous advantages (too many to mention here); but the hurdles are too much for most to even consider starting. __THE ZERONET NEEDS CONTENT. THIS WILL HELP!__

---

@imachug i have to stress how much I appreciate your detailed feedback. this has been bothering me for a while now, and I've been trying to explore many options without resolve. in the end, i'd like to move on with a solution that best suits the needs of the community, so I'd definitely appreciate your continued feedback.

cheers!
d14na

**Comment by ghost** [comment]
> yes, this is my point. this works fine, however, it doesn't work cross-zite (for the reason you stated in your first response).

Ok, wrong... for the reason he later stated, which is that public keys are the same for every zite if you use an id provider. The "auth_address" given by the "No Certificate" option is never used to sign user data for zites... for that you need to be logged into an id provider. **More importantly, you can very easily require a zite to only allow an id of an id provider that publicly stores public keys.**

Next, what you want is **already** possible. Both KxoId and ZeroId saves user public addresses in a public database.

Finally, the supposed "hurdles" haven't been explained yet, so *as far as I can tell*, there really aren't any big hurdles to what you want.

**Comment by purplesyringa** [comment]
As far as I see what you're proposing, the only thing ZeroNet doesn't support is global public keys. Instead, we have global *addresses* which are smaller and can be used for signing and verification as well. Can you use those? I.e.:

- A user comes to your ID provider 
- ZeroNet generates a unique-to-site key pair for them automatically
- The user presses the big register button
- Your server signs the user *address* with ID provider private key
- Your server adds the *address* to the database
- The user saves the signature and then can use it

The only problem I can think of is encryption (you can't encrypt with an address, a public key is required). However, you can just make the frontend generate that key and then sign its public counterpart with your auth address, or something like that.

**Comment by d14na** [comment]
> Next, what you want is already possible. Both KxoId and ZeroId saves user public addresses in a public database.

@krixano thanks for all your feedback and suggestions. however,  I'm specifically looking for the public **keys**, not **addresses**, so that a challenge can be issued that requires the user's private key to solve. now, it **IS** possible to send a challenge that validates against the public address, however, I don't see any way for the user to do validate it from the api; it requires using the cli, and that's what I'm trying to avoid.

---

actually have a solution that works (basically writing the user's public key to zite storage and reading it from the server), but its really slow (could take up to a min). it does support the current functionality without the use of plugins, so i'll use it for now, hopefully with some optimization soon to come, and perhaps _eventually_ expose it as an api for other devs looking to do the same thing.

**Comment by d14na** [comment]
> As far as I see what you're proposing, the only thing ZeroNet doesn't support is global public keys.

@imachug yes, that would be nice. its one of the main purposes for the Nametag cert provider, the other being to provide all users with an ethereum address for doing "useful" crypto (beyond just currency -- ie smart contracts).

so, your solution describes the **initial** user registration, but that's not really the problem (at least not yet). its how to "easily/conveniently" authorize a user that's visiting your zite. what i'm suggesting doesn't really fit the anonymous/pseudonymous user group, but there are lots of users who want to use zeronet and don't give a rats ass about anonymity only convenience, and this solution is for them. if anyone has used Metamask, you know that you still have to approve any dapp that wants to know who you are, but its super convenient, no usernames, no passwords. **I see this solution as a bridge from Clearnet services to migrate to Zeronet, but AT THEIR OWN PACE.**

**Comment by krixano** [comment]
>  however, I don't see any way for the user to do validate it from the api; it requires using the cli, and that's what I'm trying to avoid.

Ok, so why aren't you just asking for the ability to do this via the api? Why add more complication when you can already do what you want, you just need an api point for it?
